client/src/main/java/io/pravega/client/segment/impl/SegmentOutputStreamImpl.java

```diff
@@ -272,7 +276,31 @@ public void wrongHost(WrongHost wrongHost) {
 
         @Override
         public void segmentIsSealed(SegmentIsSealed segmentIsSealed) {
-            state.failConnection(new SegmentSealedException());
+            log.trace("Received SegmentSealed {}", segmentIsSealed);
+            if (state.segmentSealed.compareAndSet(false, true)) {
+                CompletableFuture.<Void>supplyAsync(() -> {
+                    /*
+                      Invariants for segment sealed:
+                      * SegmentSealed callback and write will not run concurrently.
```

Comment:
I don't see anything enforcing this invariant. ReplyProcessor.segmentIsSealed is invoked by the network thread when a command comes in from the wire. There is no lock here (nor should there be) so this can run in parallel to a call from the user's thread to write(). So if this thread runs first it could set segementSealed and then run the callback in the background executor, but before the callback completes write comes in and adds something new to inflight. As failConnection has not been invoked yet, it will succeed. But because getAckedEvents() was already called the event will never be seen by the callback.